src/OrchardCore/OrchardCore.Redis.Abstractions/RedisDatabaseFactory.cs

```diff
@@ -0,0 +1,49 @@
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Logging;
+using StackExchange.Redis;
+
+namespace OrchardCore.Redis
+{
+    /// <summary>
+    /// Host level factory allowing to share a <see cref="IDatabase"/> across tenants.
+    /// </summary>
+    public sealed class RedisDatabaseFactory : IRedisDatabaseFactory, IDisposable, IAsyncDisposable
+    {
+        private readonly ILogger _logger;
+        private readonly SemaphoreSlim _semaphore = new(1);
+        private IDatabase _database;
+
+        public RedisDatabaseFactory(ILogger<RedisDatabaseFactory> logger) => _logger = logger;
+
+        public async Task<IDatabase> CreateAsync(ConfigurationOptions options)
+        {
+            if (_database != null)
+            {
+                return _database;
+            }
+
+            await _semaphore.WaitAsync();
+            try
+            {
+                _database ??= (await ConnectionMultiplexer.ConnectAsync(options)).GetDatabase();
```

Comment:
I saw your comment about assuming all tenants use the same redis connection string. I wouldn't make this assumption and just create a dictionary here.